### PR TITLE
Use snippets.nvim to handle LSP snippets

### PR DESCRIPTION
**Pull request by runiq**
I *think* snippets.nvim supports pretty much all of the LSP snippets syntax.

### CODE REVIEW

**Comment by runiq** [comment]
Sorry for the noise, I forgot I already opened a PR for this branch and kept pushing WIP stuff to it. :( I have converted this to a draft PR for now.

**Comment by clason** [comment]
@runiq Would it make sense to wrap the LSP snippets in a `match_indentation` by default? (I can't think of a reason you _wouldn't_ want that.) The built-in function doesn't work for that, but you should be able to use the code fragment in the wiki.

**Comment by runiq** [comment]
@clason @mfussenegger: Actually, now that I reread [the spec](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion), I believe it might be prudent to add a `match_indentation` if the `textDocument/completion` response contains an `insertText` snippet instead of a `textEdit` snippet. The spec says that editors are allowed to mess with an `insertText` after it's been inserted, while `textEdit`s are supposed to be applied as-is.

Thoughts?

**Comment by hrsh7th** [comment]
I think VSCode does indentation in both cases.

VSCode always applies accepting completion as a snippet.
https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/editor/contrib/suggest/suggestController.ts#L358

(If the item is not `insertTextFormat != 'snippet`, VSCode does escape text and then insert as snippet)

**Comment by tjdevries** [comment]
I haven't looked super closely at this, but I think it might make sense to instead have a PR that allows for "subscribing" to the completion items or something similar, and then you could have a separate plugin (`completion-snippets.nvim` or similar) that adds this so it doesn't have to be in this plugin.

Does that seem reasonable? I could maybe do that at some point if @haorenW1025 is too busy these days.

**Comment by mfussenegger** [comment]
> @clason @mfussenegger, regarding [this](https://gitter.im/neovim/neovim?at=5f3eb035750a274130313bb3): Actually, now that I reread [the spec](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion), I believe it might be prudent to add a `match_indentation` if the `textDocument/completion` response contains an `insertText` snippet instead of a `textEdit` snippet. The spec says that editors are allowed to mess with an `insertText` after it's been inserted, while `textEdit`s are supposed to be applied as-is.
> 
> Thoughts?

I'm not sure. In my use-case it's [eclipse.jdt.ls](https://github.com/eclipse/eclipse.jdt.ls) which uses `textEdit` and the snippets don't contain the right indentation by themselves, here an example where the code should be indented:

```
newText = "@Override\nprotected ClusterBlockException checkBlock(CloseTableRequest request, ClusterState state) {\n\t${0:// TODO Auto-generated method stub\n\treturn super.checkBlock(request, state);}\n}",
```


<details>
  <summary>demo usage</summary>

  (This is not using this plugin, but my own variant and it's currently still using UltiSnips)

asciinema-2fdd309924b97885asciicast
</details>

**Comment by tjdevries** [comment]
I think `completion for snippets.nvim` and `snippets.nvim as LSP snippet manager` should probably be split into two PRs.

What do you think @runiq ?

**Comment by runiq** [comment]
Sure, will do.

**Comment by runiq** [comment]
@tjdevries The non-LSP stuff is now in #180.

**Comment by haorenW1025** [comment]
Sorry for the long wait and thanks @runiq for this PR and others for keeping the discussion going. I'm quite busy for the past several weeks but I'm getting back into `completion-nvim` now:). @runiq to my understanding this PR now only contains the snippets part of snippets.nvim (just like ultisnips and neosnippets) but not handling the LSP snippets expansion, is that correct? And is this ready for merge? It looks good to me.

**Comment by runiq** [comment]
Glad to have you back :)

Yes, exactly, only LSP handling. I think it's ready for review, at least.

File: lua/completion.lua [hunk 1]
```diff
@@ -101,6 +101,10 @@ local function hasConfirmedCompletion()
   if opt.get_option('enable_auto_paren') == 1 then
     autoAddParens(completed_item)
   end
+  if completed_item.user_data.lsp ~= nil
+	  and vim.g.completion_enable_snippet == "snippets.nvim" then
```
Comment (haorenW1025) [inline lua/completion.lua:105]: I think keep the if statement together with the applyAdditionalTextEdits is slightly better. For ex
```
  if completed_item.user_data.lsp ~= nil then
    applyAddtionalTextEdits(completed_item)
    if vim.g.completion_enable_snippet = "snippet.nvim" then
       ...
    end
  end
```

Comment (runiq) [reply lua/completion.lua:105]: Makes sense, I'll change it!

**Comment by haorenW1025** [comment]
I've tested it and it seems good for me. However I kind of like the LSP snippets to work "out of the box" instead of setting `vim.g.completion_enable_snippet`. Before we only have `vim-vsnip` as a available method so it's easier but now we should refactor bunch of stuff if we want it that way... I'll merge it first and we can refactor that later on. Thanks again @runiq !

**Comment by clason** [comment]
> However I kind of like the LSP snippets to work "out of the box" instead of setting `vim.g.completion_enable_snippet`. Before we only have `vim-vsnip` as a available method so it's easier but now we should refactor bunch of stuff if we want it that way...

I agree that the correct approach would be to automatically enable snippets (and set the corresponding client capability) if a supported snippet plugin is detected.

**Pull request merged**